[Security Solution] Add is one of operator for usage with DataProvider's QueryMatch value.

jamster10 opened the pull request:
# DISREGARD REVIEW REQUEST. commit error 🤔 😞 . Working on fix.

## Summary

This PR adds support for an is one of operator allowing users to filter multiple values for one field.

[Some investigation ](https://discuss.elastic.co/t/passing-multiple-values-in-kibana-add-filter-is-one-of/232694/2)by @andrew-goldstein revealed that since the underlying engine uses Lucene, we can add support for multiple values by using an OR query: 

`kibana.alert.workflow_status: ("open" OR "closed" OR "acknowledged")` is equivalent to 
```
"terms": {
      "kibana.alert.workflow_status": [ "open", "closed", "acknowledged"]
    }
```
Where the former is usable in our `DataProviders` used by timeline and other components that navigate a user to a pre-populated timeline.

As an enhancement to the timeline view, users can also use this `is one of` operator by interacting with the `Add field` button and selecting the new operator.

<img width="433" alt="image" src="https://user-images.githubusercontent.com/28942857/193487154-769005b6-3e5a-40bf-9476-8dd3f3bcb8ee.png">

### Checklist

Delete any items that are not applicable to this PR.

- [X] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [X] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios


## Known issues
This operator does not support timeline templates at this time so usage there disables the ability for conversion to template field but a better approach should be implemented to notify users. https://github.com/elastic/kibana/issues/142437


## Testing
Create a new timeline or visit an existing one. 
Click 'Add field' button on Timeline in OR query section
add any field ( preferably one that can have many values- consider `kibana.alerts.workflow_status` but this requires alerts.
Select the `is one of` or `is not one of operator`
Add or remove values in the value section.
Click save.

Alternatively, click an existing query, and choose Edit Filter

Commit:
Add is one of operator. and update tests

elasticmachine commented:
Pinging @elastic/security-threat-hunting (Team:Threat Hunting)

Commit:
Prevent usage of is one of operator on timeline templates

Commit:
Merge branch 'main' into 142435-add-is-one-of-operator-for-data-providers

Commit:
Fix some erroneous typings

Commit:
fix linting errors

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

Commit:
fix failing tests and snapshots

jamster10 commented on x-pack/plugins/security_solution/public/timelines/components/timeline/data_providers/add_data_provider_popover.tsx:
```diff
@@ -71,14 +71,14 @@ const AddDataProviderPopoverComponent: React.FC<AddDataProviderPopoverProps> = (
         providerToAdd: {
           id: providerId,
-          name: value,
+          name: field,
```
The name field expects a string, and value can now be of type Array, causing an error with persistence.

jamster10 commented on x-pack/plugins/security_solution/public/timelines/components/timeline/data_providers/helpers.tsx:
```diff
@@ -342,3 +342,18 @@ export const addContentToTimeline = ({
   }
 };
+
+export const convertIsOneOfQueryStringArrayToDisplayValue = (
```
The display value could be more precise than the current implementation. However since I don't think it needs to be, and this implementation is less expensive than a similar regex one, I opted for it.

jamster10 commented on x-pack/plugins/timelines/public/components/t_grid/helpers.tsx:
```diff
@@ -388,3 +406,66 @@ export const getDefaultViewSelection = ({
+  return `${field} : ${JSON.stringify(formattedValues)
+    .replace(/^\[/, '(')
+    .replace(/\]$/g, ')')
+    .replace(/,/g, ' OR ')}`;
```
Looking at this in review I think the implementation needs to be updated as if the users query includes a comma, it would be problematic. I think I'll update this to be an array.join, so all values are kept, but then wrap each value with quotation marks

andrew-goldstein commented on x-pack/plugins/security_solution/public/timelines/components/timeline/data_providers/provider.tsx:
```diff
@@ -32,7 +37,8 @@ export const Provider = React.memo<OwnProps>(({ dataProvider }) => {
       toggleExcludedProvider={noop}
       toggleEnabledProvider={noop}
       toggleTypeProvider={noop}
-      val={dataProvider.queryMatch.displayValue || dataProvider.queryMatch.value}
```
This (original, pre-PR) code appears to be incorrect, because a `displayValue` may be very different than the value that should be used to execute the query

jamster10 replied:
I agree, but was concerned this was done as some sort of bug fix. Will set it to enfoce always being value

andrew-goldstein commented on x-pack/plugins/security_solution/public/timelines/components/timeline/data_providers/provider.tsx:
```diff
@@ -32,7 +37,8 @@ export const Provider = React.memo<OwnProps>(({ dataProvider }) => {
       toggleEnabledProvider={noop}
       toggleTypeProvider={noop}
-      val={dataProvider.queryMatch.displayValue || dataProvider.queryMatch.value}
+      displayValue={String(dataProvider.queryMatch.displayValue || dataProvider.queryMatch.value)}
```
This change looks good

andrew-goldstein replied:
nit: consider using the nullish coalescing operator (`??`) here

jamster10 replied:
changed

andrew-goldstein commented on x-pack/plugins/security_solution/public/timelines/components/timeline/data_providers/provider.tsx:
```diff
@@ -32,7 +37,8 @@ export const Provider = React.memo<OwnProps>(({ dataProvider }) => {
       toggleTypeProvider={noop}
-      val={dataProvider.queryMatch.displayValue || dataProvider.queryMatch.value}
+      displayValue={String(dataProvider.queryMatch.displayValue || dataProvider.queryMatch.value)}
+      val={value}
```
Consider always using `dataProvider.queryMatch.value`, e.g.

```ts
      val={dataProvider.queryMatch.value}
```

jamster10 replied:
noted.

andrew-goldstein commented on x-pack/plugins/security_solution/public/timelines/components/timeline/data_providers/providers.tsx:
```diff
@@ -111,7 +111,9 @@ const ParensContainer = styled(EuiFlexItem)`
 `;
 
 const getDataProviderValue = (dataProvider: DataProvidersAnd) =>
```
consider removing this conditional assignment (similar to https://github.com/elastic/kibana/pull/142436#discussion_r986208635)

jamster10 replied:
fixed

andrew-goldstein commented on x-pack/plugins/security_solution/public/timelines/components/timeline/data_providers/providers.tsx:
```diff
@@ -302,6 +308,7 @@ export const DataProvidersGroupItem = React.memo<DataProvidersGroupItem>(
                 toggleExcludedProvider={handleToggleExcludedProvider}
                 toggleTypeProvider={handleToggleTypeProvider}
+                displayValue={displayValue}
                 val={getDataProviderValue(dataProvider)}
```
consider always using `dataProvider.queryMatch.value` here (similar to https://github.com/elastic/kibana/pull/142436#discussion_r986210366)

jamster10 replied:
changed

kqualters-elastic commented on x-pack/plugins/security_solution/public/timelines/components/timeline/data_providers/data_provider.ts:
```diff
@@ -13,8 +13,11 @@ export const IS_OPERATOR = ':';
+
 /** The operator applied to a field */
-export type QueryOperator = ':' | ':*';
+export type QueryOperator = ':' | ':*' | 'includes';
```
this should be `export type QueryOperator = IS_OPERATOR | EXISTS_OPERATOR | IS_ONE_OF_OPERATOR;` i think

jamster10 replied:
Tried making that change and got:
`'IS_ONE_OF_OPERATOR' refers to a value, but is being used as a type here. Did you mean 'typeof IS_ONE_OF_OPERATOR'?`

changing it to be `typeof ...` ended up being equivalent to `type QueryOperator = ':' | ':*' | 'includes';`

janmonschke replied:
```typescript
export type QueryOperator = typeof IS_OPERATOR | typeof EXISTS_OPERATOR | typeof IS_ONE_OF_OPERATOR;
```

Works when you add a couple of `typeof`'s :)

janmonschke replied:
*(deleted my previous comment)*

FWIW: I think it still makes sense going the extra mile and adding the `typeof`s :)

jamster10 replied:
got it, will do, but can I ask why

kqualters-elastic commented on x-pack/plugins/security_solution/public/timelines/components/timeline/helpers.tsx:
```diff
@@ -86,27 +91,40 @@ const checkIfFieldTypeIsNested = (field: string, browserFields: BrowserFields) =
+
+  switch (operator) {
+    case IS_OPERATOR:
+      if (!isStringOrNumberArray(value)) {
```
why no else clause here? what happens if this condition is false?

jamster10 replied:
because I reallly hope that cant happen. Not a good enough reason though :D .Would returning a dud query string work? for eg if it is an array return something like `field: value[0]`

kqualters-elastic commented on x-pack/plugins/security_solution/public/timelines/components/timeline/helpers.tsx:
```diff
@@ -86,27 +91,40 @@ const checkIfFieldTypeIsNested = (field: string, browserFields: BrowserFields) =
+        queryString + buildEXISTSQueryMatch({ browserFields, field, isFieldTypeNested });
+      break;
+    case IS_ONE_OF_OPERATOR:
+      if (isStringOrNumberArray(value)) {
```
same question as above

jamster10 replied:
Same 'dud' query? - in this case `field: value` ?

kqualters-elastic commented on x-pack/plugins/security_solution/public/overview/components/detection_response/hooks/use_navigate_to_timeline.tsx:
```diff
@@ -63,10 +65,17 @@ export const useNavigateToTimeline = () => {
-      const dataProvider = getDataProvider(mainFilter.field, '', mainFilter.value);
+      const dataProvider = getDataProvider(
+        mainFilter.field,
+        uuid.v4(),
```
why the uuid?

jamster10 replied:
needed something unique for each data provider. Without unique IDs the UI acts in a weird behaviour. I considered using the variables I already had access to to make something unique, but was worried that a user could technically make the same query or have the same filters multiple times.- A technicality to said technicality :P is that this is all in an internal hook and with correct implementation should not be a problem. 🤔 Should I change it?

michaelolo24 replied:
That's fair, we have used a combination of field values and context information for the unique id. The only way to generate something unique here is if you included the type operator, but even then a user could accidentally duplicate a field with the same operator and run into that issue. I wonder if we could use the hooks `useId` [ref](https://reactjs.org/docs/hooks-reference.html#useid) here, and also potentially use that in place of all the places were we make really contrived dataProvider id's?

jamster10 replied:
what the heck! I didn't know React gave us that outta the box. Thanks, changed!

jamster10 replied:
No wait...thats as of version 18 :P We're still using 17 :(

jamster10 commented on x-pack/plugins/security_solution/public/common/components/hover_actions/utils.ts:
```diff
@@ -18,3 +18,11 @@ export const getAdditionalScreenReaderOnlyContext = ({
   return Array.isArray(value) ? `${field} ${value.join(' ')}` : `${field} ${value}`;
 };
+
+export const convertNumericArrayToStringArray = (
```
Opted to memoize the values and within the useMemo do type checks. I think initially I needed that function in more than one place, but no longer. Let me know if that's not an ok implementation though.

```
 const values = useMemo(() => {
    const val = dataProvider.queryMatch.value;

    if (typeof val !== 'number') {
      if (Array.isArray(val)) {
        return val.map((item) => String(item));
      }
      return val;
    }
    return `${dataProvider.queryMatch.value}`;
  }, [dataProvider.queryMatch.value]);
```

It's guaranteed to be a string or string[ ]

logeekal replied:
a minor nit-pick. Do you think below is more understandable? Please ignore if you think wise.

```
if(typeof val === 'number' ) return `${val}`

if(Array.isArray()){
     return val.map((item) => String(item));
}

return val

```

jamster10 replied:
agreed & changed!

jamster10 commented on x-pack/plugins/security_solution/public/timelines/components/timeline/data_providers/provider.tsx:
```diff
@@ -19,6 +19,11 @@ interface OwnProps {
 export const Provider = React.memo<OwnProps>(({ dataProvider }) => {
   const [isPopoverOpen, setIsPopoverOpen] = useState(false);
 
+  const value =
```
cool! was worried this was intentional for bug reasons I didnt know.

jamster10 commented on x-pack/plugins/security_solution/public/timelines/components/timeline/data_providers/providers.tsx:
```diff
@@ -264,6 +266,10 @@ export const DataProvidersGroupItem = React.memo<DataProvidersGroupItem>(
     );
 
+    const displayValue = String(
+      dataProvider.queryMatch.displayValue || dataProvider.queryMatch.value
```
changed

michaelolo24 commented on x-pack/plugins/security_solution/public/timelines/components/timeline/helpers.tsx:
```diff
@@ -253,3 +271,71 @@ export const focusUtilityBarAction = (containerElement: HTMLElement | null) => {
+  value: Array<string | number>;
+}): string => {
+  if (isFieldTypeNested) {
+    // some kind of nested handler implementation if needed, I assume with BrowserFields?
```
Was this something you planned to implement for this pr?

jamster10 replied:
Its a conversation I have set up with Xavier for today to see if it's necessary. For now I am scraping it out though.

michaelolo24 commented on x-pack/plugins/security_solution/public/timelines/components/timeline/helpers.tsx:
```diff
@@ -253,3 +271,71 @@ export const focusUtilityBarAction = (containerElement: HTMLElement | null) => {
+  val: string | number | Array<string | number>
+): val is Array<string | number> {
+  return (
+    typeof val !== 'string' &&
```
You could probably simplify this with `Array.isArray(val) && (typeof val[0] === 'string' || typeof val[0] === 'number')`

jamster10 replied:
changed

michaelolo24 commented on x-pack/plugins/timelines/common/types/timeline/data_provider/index.ts:
```diff
@@ -13,8 +13,11 @@ export const IS_OPERATOR = ':';
+
 /** The operator applied to a field */
-export type QueryOperator = ':' | ':*';
+export type QueryOperator = ':' | ':*' | 'includes';
```
Don't forget to update this type if you end up updating the other one above

jamster10 replied:
changed!

michaelolo24 commented on x-pack/plugins/timelines/public/components/t_grid/helpers.tsx:
```diff
@@ -388,3 +406,66 @@ export const getDefaultViewSelection = ({
+  val: string | number | Array<string | number>
+): val is Array<string | number> {
+  return (
+    typeof val !== 'string' &&
```
Same thing here, you can just check first if `Array.isArray(value) instead of the val string and number checks` Then you can re-use this function above.

Maybe an over-engineered solution could be:

export IsArrayOfOptions = 'string' | 'number' | 'object' .... etc

export const isArrayOf = (value: unknown, options: IsArrayOfOptions[]) =>  Array.isArray(value) && options.some((typeOption) => typeof value[0] === typeOption);

isArrayOf(val, ['number', 'string']);

jamster10 commented on x-pack/plugins/security_solution/public/common/components/event_details/table/use_action_cell_data_provider.ts:
```diff
@@ -58,7 +64,8 @@ export const getDataProvider = (field: string, id: string, value: string): DataP
     value,
-    operator: IS_OPERATOR,
+    operator,
+    displayValue: convertIsOneOfQueryStringArrayToDisplayValue(value),
```
changed the name to getDisplayValue instead since it can be used anywhere.

Commit:
further changes from code review

Commit:
Merge branch '142435-add-is-one-of-operator-for-data-providers' of github.com:jamster10/kibana into 142435-add-is-one-of-operator-for-data-providers

jamster10 commented:
> Thanks @jamster10 for this awesome feature, which will serve as a building block for several new features / fixes!
> 
> In this first round of desk testing, I noticed the behavior the `is NOT` operator has unexpected results. For example, the `NOT kibana.alert.severity: "high"` in the screenshot below produces the expected result in `main`, but unexpected results in the PR branch, per the screenshot below:
> 
> ![not-high-severity](https://user-images.githubusercontent.com/4459398/193655117-5579954c-2b1a-4a39-8788-5a310fb3a159.png)
> 
> The screenshot of a JSON diff below illustrates the difference between the request in `main` (left) and the PR branch (right):
> 
> ![json-diff](https://user-images.githubusercontent.com/4459398/193655738-5ec2cba9-8f5b-4480-8927-17ad1e9e347f.png)

Thanks for catching this Andrew, I will update this conversation when I have resolved it and ping you and your team

andrew-goldstein commented on x-pack/plugins/security_solution/public/timelines/components/edit_data_provider/index.tsx:
```diff
@@ -277,3 +303,127 @@ export const StatefulEditDataProvider = React.memo<Props>(
 
 StatefulEditDataProvider.displayName = 'StatefulEditDataProvider';
+
+function isStringOrNumberArray(
```
Given the following unit tests:

```ts
  describe('isStringOrNumberArray', () => {
    test('it returns false when value is not an array', () => {
      expect(isStringOrNumberArray('just a string')).toBe(false);
    });

    test('it returns false when value is an array of mixed types', () => {
      expect(isStringOrNumberArray(['mixed', 123, 'types', {}, undefined, null])).toBe(false);
    });

    test('it returns true when value is an empty array', () => {
      expect(isStringOrNumberArray([])).toBe(true);
    });

    test('it returns true when value is an array of all strings', () => {
      expect(isStringOrNumberArray(['all', 'string', 'values'])).toBe(true);
    });

    test('it returns true when value is an array of all numbers', () => {
      expect(isStringOrNumberArray([123, 456, 789])).toBe(true);
    });
  });
```

The current implementation fails some tests:

```
    isStringOrNumberArray
      ✓ it returns false when value is not an array
      ✕ it returns false when value is an array of mixed types (1 ms)
      ✕ it returns true when value is an empty array
      ✓ it returns true when value is an array of all strings
      ✓ it returns true when value is an array of all numbers
```

Consider something like the following changes to address the test failures:

```ts
export const isStringOrNumberArray = (value: unknown): value is Array<string | number> =>
  Array.isArray(value) &&
  (value.length === 0 ||
    value.every((x) => typeof x === 'string') ||
    value.every((x) => typeof x === 'number'));
```

Would you be willing to adopt the unit tests above?

jamster10 replied:
I have but left out the second check:
 value.every returns true for an empty array:
 https://stackoverflow.com/a/34137289

andrew-goldstein commented on x-pack/plugins/security_solution/public/timelines/components/edit_data_provider/index.tsx:
```diff
@@ -277,3 +303,127 @@ export const StatefulEditDataProvider = React.memo<Props>(
+  return Array.isArray(val) && (typeof val[0] === 'string' || typeof val[0] === 'number');
+}
+
+const isValueFieldInvalid = (type: DataProviderType, value: string | number) =>
```
consider extracting this function out of the component and adding unit tests for it

jamster10 replied:
done!

andrew-goldstein commented on x-pack/plugins/security_solution/public/timelines/components/edit_data_provider/index.tsx:
```diff
@@ -277,3 +303,127 @@ export const StatefulEditDataProvider = React.memo<Props>(
+  );
+};
+
+const convertValuesToComboboxValueArray = (
```
Consider extracting this function, and similar ones below like `convertComboboxValuesToStringArray` out of this component, because these pure functions can be unit tested independently from the component.

jamster10 replied:
done!

andrew-goldstein commented on x-pack/plugins/security_solution/public/timelines/components/edit_data_provider/index.tsx:
```diff
@@ -277,3 +303,127 @@ export const StatefulEditDataProvider = React.memo<Props>(
+  return values.map((item) => item.label);
+};
+
+const ControlledComboboxInput = ({
```
Consider moving this component into its own file, so it can be unit tested independently.

jamster10 replied:
done!

andrew-goldstein commented on x-pack/plugins/security_solution/public/timelines/components/flyout/header/index.tsx:
```diff
@@ -145,6 +146,8 @@ const FlyoutHeaderPanelComponent: React.FC<FlyoutHeaderPanelProps> = ({ timeline
 
   const { euiTheme } = useEuiTheme();
 
+  console.log({ isDataInTimeline, test: combinedQueries?.filterQuery });
```
nit: consider removing this if it's no longer required

jamster10 replied:
sorted.

andrew-goldstein commented on x-pack/plugins/security_solution/public/timelines/components/timeline/data_providers/helpers.test.tsx:
```diff
@@ -1103,4 +1104,10 @@ describe('helpers', () => {
+
+  describe('convertIsOneOfQueryStringArrayToDisplayValue', () => {
+    it('converts a "is one of" query to correct format for a string array', () => {
+      expect(convertIsOneOfQueryStringArrayToDisplayValue(['a', 2, 'c'])).toBe('( a OR 2 OR c )');
```
Consider removing the extra whitespace after the opening `(`, and before the closing `)`:

**Before**
`( a OR 2 OR c )`

**After (suggested)**
`(a OR 2 OR c)`

jamster10 replied:
changed 👌

andrew-goldstein commented on x-pack/plugins/security_solution/public/timelines/components/timeline/data_providers/helpers.tsx:
```diff
@@ -342,3 +342,18 @@ export const addContentToTimeline = ({
   }
 };
+
+export const getDisplayValue = (
```
consider adding tests for this function

andrew-goldstein commented on x-pack/plugins/security_solution/public/timelines/components/timeline/data_providers/helpers.tsx:
```diff
@@ -342,3 +342,18 @@ export const addContentToTimeline = ({
+  return value;
+};
+
+function isStringOrNumberArray(
```
consider applying the feedback [in this comment](https://github.com/elastic/kibana/pull/142436#discussion_r987095667) here

andrew-goldstein commented on x-pack/plugins/security_solution/public/timelines/components/timeline/data_providers/provider_item_badge.tsx:
```diff
@@ -143,6 +145,7 @@ export const ProviderItemBadge = React.memo<ProviderItemBadgeProps>(
           providerId={providerId}
           togglePopover={togglePopover}
           toggleType={onToggleTypeProvider}
+          displayValue={displayValue || String(val)}
```
nit: consider replacing `||` with `??`

andrew-goldstein commented on x-pack/plugins/security_solution/public/timelines/components/timeline/data_providers/provider_item_actions.tsx:
```diff
@@ -138,7 +138,7 @@ export const getProviderActions = ({
         ? {
             className: CONVERT_TO_FIELD_CLASS_NAME,
-            disabled: isLoading,
+            disabled: isLoading || operator === IS_ONE_OF_OPERATOR,
```
Thanks for disabling the option to covert the `IS_ONE_OF` operator to a template field.

Everything in the screenshot below is working as expecetd:

![working-as-expected](https://user-images.githubusercontent.com/4459398/193941852-8f166280-c7e8-43d9-aa5c-0a3912c95bc0.png)

The `Add field` case, shown below is also working as expected:

![add-field-timeline-template](https://user-images.githubusercontent.com/4459398/193942007-9174525c-3647-4d26-bf7c-94d64d07da37.png)

There are a two additional places in the UI where similar checks are necessary (to disable the option to convert the field to a template field):

1) The `is one of` and `is not one of` operators should not be shown in the `Add template field` scenario, shown in the screenshot below:

![add-template-field-timeline-template](https://user-images.githubusercontent.com/4459398/193942288-adc77d88-4cb8-4513-8e80-e30331344bae.png)

2) The `Convert to template field` option in the popover menu below should not be shown when the operator is `is one of`, or `is not one of`, per the screenshot below:

![popover-menu](https://user-images.githubusercontent.com/4459398/193942463-398fb931-2e01-4178-8ad8-17c16fcf14d5.png)

jamster10 replied:
fixed!

andrew-goldstein commented on x-pack/plugins/timelines/public/components/t_grid/helpers.tsx:
```diff
@@ -388,3 +406,60 @@ export const getDefaultViewSelection = ({
 /** This local storage key stores the `Grid / Event rendered view` selection */
 export const ALERTS_TABLE_VIEW_SELECTION_KEY = 'securitySolution.alerts.table.view-selection';
+
+const buildISQueryMatch = ({
```
consider adding unit tests for this function, and the `buildEXISTSQueryMatch` and `buildEXISTSQueryMatch` functions below

jamster10 replied:
Done!

andrew-goldstein commented on x-pack/plugins/security_solution/public/timelines/components/timeline/helpers.tsx:
```diff
@@ -253,3 +274,57 @@ export const focusUtilityBarAction = (containerElement: HTMLElement | null) => {
+    : `${field} ${EXISTS_OPERATOR}`;
+};
+
+const buildISONEOFQueryMatch = ({
```
Consider adding a unit test for the `combineQueries` function (defined in `x-pack/plugins/timelines/public/components/t_grid/helpers.tsx`) that asserts both the data provider (with an `IS_ONE_OF` operator) shown in the screenshot below, and the (KQL) filter representation (also in the screenshot below), are both converted to the same ES query DSL:

![assert-same-query-dsl](https://user-images.githubusercontent.com/4459398/193946302-62f114a3-1d41-4298-b263-3ed206a4273b.png)

andrew-goldstein commented on x-pack/plugins/timelines/public/components/t_grid/helpers.tsx:
```diff
@@ -87,27 +93,39 @@ const checkIfFieldTypeIsNested = (field: string, browserFields: BrowserFields) =
+        queryString =
+          queryString + buildISONEOFQueryMatch({ browserFields, field, isFieldTypeNested, value });
+      }
+      break;
```
Consider using something like the `assertUnreachable` utility function in `x-pack/plugins/timelines/common/utility_types.ts` to add a `default` case that ensures this `switch` is exhaustive (at compile time), per the details in https://stackoverflow.com/questions/39419170/how-do-i-check-that-a-switch-block-is-exhaustive-in-typescript

jamster10 replied:
thats pretty cool, thanks for pointing it out

logeekal reviewed (changes requested):
Since I am developing a feature which is based on this one, I was doing some testing where I observed below issue.

The edit filter popup is crossing the `left` edge of the screen. 

![Screenshot 2022-10-05 at 16 00 13](https://user-images.githubusercontent.com/7485038/194079639-39c2fa70-7043-4c32-ba4f-d825e827adb2.png)

Commit:
merge main and update useNavigateToTimeline

Commit:
further changes from code review

Commit:
merge main

Commit:
fix lint problems

Commit:
further clean up

Commit:
fixes

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

breehall committed:
Bumping EUI to version 67.1.8 (#141279)

* Updated EUI to version 67.1.2. Updated instaces of ButtonColor from EUI to EuiButtonColor.

* Updated to EuiCard instances that utilize the betaBadgeProps object to return an empty string instead of undefined when the label is unavailable

* Removed two instances of the deprecated internetExplorerOnly() mixin

* Updated two instances of the ButtonColor import to EuiButtonColor as is was renamed in PR #6150

* Updated snapshots in Jest Test Suite #1 to account for EuiButton and EuiCard Emotion conversions. Updated snapshots for EuiTooltip as it now contains the new EuiToolTipAnchor component that replaced the tooltip anchor styles

* Updated snapshots in Jest Test Suite #2 to account forEuiButton, EuiDescriptionList, EuiButtonIcon, and EuiBadge Emotion conversions.

* Updated snapshots in Jest Test Suite #3 to account for EuiDescriptionList, EuiButton, and EuiBadge Emotion conversions. Updated snapshots for EuiTooltip as if now contains the new EuiTooltipAnchor component that replaced the tooltop anchor styles

* Updated snapshots in Jest Test Suite #4 to account for EuiButton Emotion conversion.

* Updated snapshots in Jest Test Suite #5 to account for EuiButton Emotion conversion.

* Updated snapshots in Jest Test Suite #8 to account for EuiButtonIcon and EuiButton Emotion conversions. Updated snapshots for EuiTooltip as it now contains the new EuiTooltipAnchor component that replaced the tooltip anchor styles.

* Updated snapshots in Jest Test Suite #9 to account for EuiFlyout and EuiButton Emotion conversions.

* Updated snapshots in Jest Test Suite #10 to account for EuiButton, EuiBadge, EuiButtonIcon, and EuiCard Emotion conversions. Updated snapshots for EuiToolTtip as it now contains the new EuiTooltipAnchor component that replaced the tooltip anchor styles

* Updated instances of EuiButtonIconColor to use EuiButtonIconProps['color'] as it was removed in PR #6150

* Updated tests that target EuiButton to simulate click events to target a generic button to prevent undefined click event errors

* Updated snapshots in Jest Test Suite #1 to account for EuiButton and EuiCard Emotion conversions

* Added the EuiFlyout mixins and variables to Lens Sass file as EuiFlyout has been converted to Emotion and the Sass styles are no longer available in EUI

* Added the EuiCallOutTypes variable to Step Progress Sass file as EuiCallOut has been converted to Emotion and the Sass styles are no longer available in EUI

* Updated snapshots in Jest Test Suite #2 to account for recent Emotion conversions.
Updated snapshots in server_status.test.tsx to render EuiBadge before checking the snapshots to reduce the snapshot churn caused by Emotion.
Updated tests that target EuiButton to simulate click events to target a generic button to prevent undefined click event errors

* [CI] Auto-commit changed files from 'node scripts/precommit_hook.js --ref HEAD~1..HEAD --fix'

* Added imports for the added flyout mixin. Removed references to EuiCallOut mixin as the component has been converted to Emotion and is no longer available for use.

* Updated unit tests and snapshots in Jest Test Suite #10.
Updated snaphshots to account for EuiBadge, EuiDescriptionList, EuiFlyout, and EuiCard Emotion conversions. Updated snapshots for EuiTooltip as it now contains the new EuiTooltipAnchor component that replaced the tooltip anchor styles.
Updated tests that target EuiButton to simulate click events to target a generic button element to prevent undefined click event errors

* Updated unit tests in Jest Test Suite #11 that target EuiButton to simulate click events to target a generic button to prevent undefined click event errors

* Updated unit tests in Jest Test Suite #12 by updating tests that target EuiButton to simulate click events. Instead, these tests now target a generic button element to prevent undefined click event errors

* Updated unit tests in Jest Test Suite #1 by updating tests that target EuiButton to simulate click events. Instead, these tests now target a generic button element to prevent undefined click event errors

* Updated unit tests in Jest Test Suite #2 by updating tests that use EuiButton to simulate click events. Instead, these test have been updated to target a button element to prevent undefined click event errors.

* [CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

* Updated reference to mixins Sass file.
Updated snapshots for Jest Test Suite #5 to account for EuiButton Emotion conversion. Updated unit tests that target EuiButton to simulate click events. These tests have been updated to target a button element to prevent undefined click event errors

* Updated unit tests in Jest Test Suites 3, 7, 8, 13, and 14.
Updated snapshot to account for EuiButton Emotion conversion.
Updated tests that target EuiButton to simulate click events. These tests now target a generic button element to prevent undefined click event errors.
Updated a few snapshots by adding .render() before checking the snapshot. This will prevent large snapshots coming from recent Emotion conversions

* Updated snapshots in Jest Test Suite #10 to account for the recent EuiButton Emotion conversion

* Updated unit tests in Jest Test Suite #2 by editing tests that target EuiButton to simulate click events. These tests now target a button element in order to prevent undefinde click event errors

* Updated snapshots in Jest Test Suite #10 to account for EuiButton and EuiDescriptionList Emotion conversions

* Updated test cases in Jest Test Suites 3, 7, and 8. Updated snapshots to account for EuiButton and EuiPagination Emotion conversions.
Updated tests that target EuiButton to simulate click events. These tests now target a button element to prevent undefined click errors

* Updated test cases in Jest Test Suite 14. Updated snapshots to account for EuiButton Emotion conversion. Opted to use .render() when updating a few snapshots to reduce the large length of snapshots caused by Emotion

* [CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

* Revised a change to betaBadgeProps to ensure that the label is available. If not, the value for the badge with be set to undefined.

* Resolved two linting errors

* Resolved two linting errors

* Updated Jest unit tests in various suites.
Updated snapshots to account for EuiButton Emotion conversion. Updated snapshots for EuiTooltip as it now contains the new EuiTooltipAnchor component that replaced the tooltip anchor styles.

* Updated EuiFlyout in query_flyout.tsx to remove the onClick function from maskProps as it is no longer available. Updated this flyout to use ownFocus and not to close when the overlay mask is clicked.

* Removed the use of EuiButtonIconColor in favor of EuiButtonIconProps['color']

* [CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

* Updated Cypress test looking for strict equality on EuiPaginationButton class names to match a substring of the Emotion generated class name

* Removed unneeded debugging code. Updated snapshots for various test suites to account for the recent EuiButton Emotion conversion

* Updated a few EuiButton, EuiButtonEmpty, and EuiText components that set the color as ghost. The ghost color mode has been deprecated as of PR #6150. These components now are wrapped in EuiThemeProvider with a dark colorMode to create the previous ghost color.

* Resolved  TS error with EuiCard betaBadgeProps

* [CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

* Remove references to now-removed EuiFlyout CSS classes/vars

* Remove now-removed euiBadge className references

- Convert directly to EuiBadge instead of using CSS

- Remove confusing and now-possibly-irrelevant CSS badge overrides - left/right icons are now set via JSX and not via flex-direction

* Pre-emptively fix various euiOverlayMask CSS overrides

- this data attr isn't technically in yet but will be once https://github.com/elastic/eui/pull/6289 merges

- at the very least this isn't breaking any more than it currently already is!

* Update to v67.1.3

* v67.1.4

* Resolved test failing test case in Security/Manage/Blocklist. The test did not remove focus from the last combo box in the form, which didn't allow the disbaled attribute to be removed from the flyout submit button. I've updated the mock file for Blocklist to return focus to the first form element in the flyout to allow the disabled attribute to be removed.

* Updated snapshots to account for the recent EuiText Emotion conversion

* Fix Log's custom tooltips relying on EuiTooltip classNames that no longer exist

* Fix Vega vis custom tooltips relying on EuiTooltip classNames that no longer exist

- this one is trickier than Log's as it's not using React, so we need to use Emotion's Global to set a static className

* Convert remaining vega_vis.scss to Emotion

- as an example of how other global + non global styles could be handled in the future

* Fix references to removed `euiPaginationButton-isActive` className

- use aria-current attribute instead

* Added missing EuiFlyoutAnimation keyframes for EuiFlyout. This resolved test that failed because they used onAnimationEnd because the FlyoutAnimation could not be found.

* Reolved Jest Tests in suites 1 and 5. Updated snapshots to account for the recent EuiButton Emotion conversion. Updated snapshots for EuiToolTip as it now contains the new EuiToolTipAnchor component that replaced the tooltip anchor styles.

* iterate on rules_list.test.tsx

* bump eui to v67.1.5

* Updatde snapshots for jest test suites to account for the recent EuiButton, EuiOverlayMask, EuiTooltip, and EuiBadge Emotion conversions

* Resolved failing security test by updating the target element for CONNECTOR_TITLE. EuiCard has recently been converted to Emotion and the card title is no longer wrapper in a span.

* Resolved failing test case in Runtime Fields. The modify runtime field test was failing because the combobox responsbible for adding and updating scripts was not appearing. The textbox did not appear because the shared setFieldScript function targets and toggles the script textbox when opening the flyout. When a runtime field is being modified, the toggle is already active and using the shared function will trigger the toggle again (losing access to the script textbox).
Also resolved an issue that prevented the warning EuiCallout to appear when changing the type of a runtime field from its original type. Resolved this by adding an enter keypress at the end of setFieldType function to confirm the type selection, thus triggering the EuiCallout

* Resolved two tests that were failing in Lens. These test were failing because they were checking for equality in class names that no longer exist within EuiButtonGroup as it was recently converted to Emotion. These tests were updated to check for a substring of the new and longer class name

* Quick fix in test case failing because of misspelling in data-test-sub

* Updated snapshot for Jest test case as EuiButton as recently been converted to Emotion

* Removed console.log statement. Oops!

* Resolved a failing test case in Lens. They were failing because they were checking for equality in class names that no longer exist within EuiButtonGroup as it was recently converted to Emotion. These tests were updated to check for a substring of the new and longer class name.
Updated a Security test case by giving a target button the data-test-subj attribute for easier querying

* Removed reference to EuiFlyout mixin as it has been converted to Emotion. Updated the reference to an interal copy of EuiFlyout styles

* Corrected spelling error in EuiFlyout animation in Lens app

* Update EUI with latest backport

* Update button snapshots

* fix another button snapshot

* More snapshot fixes

* [EuiButton][Security] Fix button relying on now-removed `euiButton__text` CSS

- replace removed CSS with `eui-textTruncate` util instead

- combine/DRY out unnecessary span - was affecting min-width of truncation util

+ increase screenshot diff limit - this was smaller than updating the actual baseline screenshots for whatever reason (likely render diff between local and CI)

* Fix remaining Jest tests affected by Emotion conversions

- because Emotion creates its own wrapper, `.first()` can no longer be used - prefer `.last()` instead

* Fix Jest test affected by EuiButton Emotion conversion + removed modifier class

- targeting the native DOM node + filtering by disabled true/false gets us back to the 'correct' lengths

* Fix + improve flyout test

- `.last()` changes to account for EuiButton Emotion conversion is needed, but the last onClose assertion still fails due to us having modified inputs, and the confirm modal being displayed

- split test into two separate tests - one testing the onClose call, and the other testing the confirm modal

* derpin

* Skip rules_list Jest suite

* Update new EuiButton snapshot

* Upgraded EUI version to 67.1.7

* [EuiCard] Update snapshots

* [EuiPopover] Update snapshots

* [QA] Fix missing Vega warn/error message colors

;_;

* [CI] Auto-commit changed files from 'node scripts/generate codeowners'

* Fix Lens kbnToolbarButton regressions

- Caused by flattening of EUI button CSS specificity

- background-color was previously relying on isDisabled CSS specificity to override its #fff color

- `text` color modifier & `!important` is no longer needed and overrides Emotion CSS flatly

- isDisabled class is no longer needed - euiButton no longer sets `pointer-events: none` on disabled buttons (fixes tooltip bug in webkit as well)

* Backport EUI 67.1.8 fixes

* Update EuiCard snapshots

* Fix EuiModal form wrapper causing overflow issues

- see https://elastic.github.io/eui/#/layout/modal#forms-in-a-modal

* Workaround for `.kbnOverlayMountWrapper` mount point causing overflow issues

- not sure what all is using this modal service to be honest, but the wrapper is causing issues with the modal layout, this fixes overflow issues but will not fix any mask-image issues as a result

* more snapshot updates

* EuiButton - added textProps to EuiButton to prevent very long button names from spilling over outside of the container

* EuiButton - Update EuiButton related snapshots. Updated tests that target EuiButton directly to use a data-telementary-id for more specific element querying required by Emotion

* QA - Removed unnecessary comment in code

* Temporary fix for EuiCard[selectable][layout=horizontal] instances on security solutions' rule page

* Temporary fix for EuiCard[selectable][layout=horizontal] instances on osquery live query and canvas's datasource selector

* [CI] Auto-commit changed files from 'node scripts/precommit_hook.js --ref HEAD~1..HEAD --fix'

* Fix CSS specificity, where canvas's solutionToolbarButton's background-color now takes precedence over EuiButton's primary styles

* Removed update to search_marker_tooltip that removed the euiTooltip styles and replaced then with Emotion styling. Added EuiTooltip Sass styles for the component to rely on to test for a styling bug that is causing the tooltip and the tooltip arrow to be out of sync with each other.

* Lint Sass file

* Lint Sass file

* Removed overflow:hidden style from .vgaVis_view as it was causing euiScrollStyles not to present the scroll bars in Vega Vis

* Remove typo from EuiButton textProps object. 'className' should not have been included in the actual class name

* Revert tooltip Sass

This reverts commit 20e6ead5713781c35fdd15c73e9886f9abcecfac, a5cd2de901caec2d485c162eed184613e6a7b32b, and c605cbd7b92311b76f05ed1d6fab0651eeeae0cb

* Fix Emotion tooltip arrows

Co-authored-by: kibanamachine <42973632+kibanamachine@users.noreply.github.com>
Co-authored-by: Constance Chen <constance.chen@elastic.co>
Co-authored-by: Chandler Prall <chandler.prall@elastic.co>

Heenawter committed:
Change wording of tooltip (#144652)

mistic committed:
skip flaky suite (#143007)

Commit:
Getting started ccs tests (#144656)

* Removed comment of the issue that was referenced for the skip. But the tests were already skipped.

* Added a describe block for the tests that involve space b. There is an accessibility issue that causes one test to fail and then the subsequent test fails. The issue has been logged. The rest of the tests have been unskipped.

* Added CCS Test for integration test.

* Added CCS config for getting started shakespeare tests.

* Removed unused reference.

* Restored tests.

Co-authored-by: cuffs <cuffs@cuffss-Office-MacBook-Pro.local>
Co-authored-by: Kibana Machine <42973632+kibanamachine@users.noreply.github.com>

mistic committed:
chore(NA): include .ijwb folder into the .gitignore file (#144755)

mistic committed:
skip flaky suite (#111821)

lcawl committed:
[DOCS] Fix links in Kibana troubleshooting (#144749)

kibanamachine committed:
[api-docs] Daily api_docs build (#144773)

Kuznietsov committed:
[Lens] XY expression types improvement. (#144327)

* referenceLineDecorationConfig type safety added.

* dataDecorationConfigFn type safety added.

* Added extendedDataLayerFn type safety.

* Added extendedAnnotationLayerFn type safety.

* Added referenceLineLayerFn type safety.

* Added yAxisConfigsToExpression type safety.

* Added more function builders.

* Added layeredXyVisFn type safety.

* Fixed tests.

Co-authored-by: Kibana Machine <42973632+kibanamachine@users.noreply.github.com>

machadoum committed:
Add default data view patterns to all timeline analyzer requests (#144705)

* Add default data view patterns to all timeline analyzer requests

Co-authored-by: Kevin Qualters <kevin.qualters@elastic.co>
Co-authored-by: kibanamachine <42973632+kibanamachine@users.noreply.github.com>
Co-authored-by: Spencer <spencer@elastic.co>

flash1293 committed:
[Lens] Fix suggestion bug (#144708)

* fix suggestion bug

* [CI] Auto-commit changed files from 'node scripts/generate codeowners'

Co-authored-by: kibanamachine <42973632+kibanamachine@users.noreply.github.com>
Co-authored-by: Stratoula Kalafateli <efstratia.kalafateli@elastic.co>

flash1293 committed:
fix bug (#144716)

Co-authored-by: Stratoula Kalafateli <efstratia.kalafateli@elastic.co>

hop-dev committed:
[Fleet] Add `?full` option to get package info endpoint to return all package fields (#144343)

* add source mode to archive parsing

* add ?full patram to get pkg info

* revert archive change

* do not cache EPR fetch info

* add test package for source_mode

* add int test

* add new param to frontent

* use released package for test

* tidy for PR

* Add API docs

* add get all packages script

* get archive from cache even if not installed

* fix types

shahzad31 committed:
[Synthetics] Added step metric stats (#144594)

yuliacech committed:
[Guided onboarding] More API fixes (#144729)

* [Guided onboarding] More fixes to the API requests logic preventing spamming the endpoint

* [Guided onboarding] Update the tests

* [Guided onboarding] Add a comment to the code

PhilippeOberti committed:
[TIP] Fix indicators table loading issue and add e2e to check existence of columns (#144371)

https://github.com/elastic/security-team/issues/5317

renovate[bot] committed:
Update dependency elastic-apm-node to ^3.40.0 (#144760)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

rockdaboot committed:
[Profiling] For diff functions add change of total samples (#144724)

suchcodemuchwow committed:
feature: report performance metrics for files plugin

tomsonpl committed:
[Osquery] [Fix] - Set snapshot as default result type (#144786)

jcger committed:
[RAM] Remove errors and warning in triggers_actions_ui jest test (#144443)

* fix action_form test

* fix rule test

* improve performance

* split huge test

* [CI] Auto-commit changed files from 'node scripts/generate codeowners'

* include skipped tests

* fix test and remove not needed optional

Co-authored-by: kibanamachine <42973632+kibanamachine@users.noreply.github.com>
Co-authored-by: Xavier Mouligneau <xavier.mouligneau@elastic.co>

TattdCodeMonkey committed:
[Enterprise Search][ML Inference] Re-use pipeline with unavailable model (#144483)

* added `ml-inference-` prepend to pipeline name text input instead of describing it in the help text

* added ability to simulate with an existing inference pipeline

* fix set processor if clause

* test: fixing unit tests for updates

* test: simulate existing pipeline

Co-authored-by: Kibana Machine <42973632+kibanamachine@users.noreply.github.com>

stratoula committed:
[Lens] Errors out used columns that don't exist on the index pattern used (#144781)

benakansara committed:
Adds new contextual attributes to Infrastructure - Metric threshold rule (#143001)

* creating alert link with alert id and rule id

* Revert "creating alert link with alert id and rule id"

This reverts commit 29bede6aabc5772d19f98e09cda73e8df4103c52.

* additional context variables for metric threshold rule

* [CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

* fixing type error

* changed description of context variables

* adding context only if group by is available, creating container list

* refactoring, fixing types

* fixing CI error

* fixed issues

* refactoring

* fixing failing tests

* refactoring, fixing tests

* added tests, refactoring

* fixing issue with group by

* fixing failing tests

* fixing failing tests

* [CI] Auto-commit changed files from 'node scripts/precommit_hook.js --ref HEAD~1..HEAD --fix'

* [CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

* [CI] Auto-commit changed files from 'node scripts/precommit_hook.js --ref HEAD~1..HEAD --fix'

* fixing failing tests

* [CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

* refactoring

* [CI] Auto-commit changed files from 'node scripts/precommit_hook.js --ref HEAD~1..HEAD --fix'

Co-authored-by: kibanamachine <42973632+kibanamachine@users.noreply.github.com>

kdelemme committed:
feat(slo): Introduce list SLOs route with search capabilities (#144561)

doakalexi committed:
Fixing flaky test (#144735)

cnasikas committed:
[Cases] Add spaceId to cases backlink (#144616)

* Add spaceId to backlink to cases

* Add integration tests

Co-authored-by: Kibana Machine <42973632+kibanamachine@users.noreply.github.com>

alexwizp committed:
[Lens] unify UI styles for warning messages (#144800)

* [Lens] update warning messages styles

* update jest

juliaElastic committed:
Refactor bulk update tags to fix issue with concurrently removing multiple tags (#143543)

* update tags refactor

* reverted yarn file

* update tags with abort on conflict and retry

* fixed hosted policy condition

* fixed unit tests

* fixed integration tests, fixed a bug

* fixed unit test

* fixed unit test

* fix cypress test

* small cleanup

alexwizp committed:
[Lens] embeddable warnings placement looks pretty bad with a metric. (#144368)

* [Lens] embeddable warnings placement looks pretty bad with a metric.

* Update warnings.tsx

Co-authored-by: Kibana Machine <42973632+kibanamachine@users.noreply.github.com>

renovate[bot] committed:
Update babel to ^7.20.1 (main) (#144798)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

ofiriro3 committed:
[Cloud Posture] Remove postCreateHook for creating the cfg data.yaml (#144006)

[Cloud Posture] Remove postCreate hook

dokmic committed:
[Saved Objects] Add documentation covering hidden saved object types (#144647)

hop-dev committed:
[Fleet] Change broken "delete key" test selector (#144805)

* change test selector

* move to const

flash1293 committed:
[TSVB] Adjust tolerance of screenshot (#144804)

Fixes https://github.com/elastic/kibana/issues/138814 by updating the
baseline and reducing the tolerance

Commit:
rename codeowners from shared-ux to global-experience (#144746)

https://github.com/elastic/kibana/pull/144326 attempted to change the
owners for packages, but because of a misconfiguration in CI the correct
codeowners file wasn't determined until after the PR was merged. This
fixed the issue by modifying the kibana.jsonc files of the packages to
use the correct owner, which then ends up in the codeowners file.

Co-authored-by: kibanamachine <42973632+kibanamachine@users.noreply.github.com>

jbudz committed:
Bump node to 16.18.1 (#144751)

https://github.com/nodejs/node/blob/main/doc/changelogs/CHANGELOG_V16.md#16.18.1

alvarezmelissa87 committed:
[ML] Data Frame Analytics: Add brushing to link docs in scatterplot charts (#144354)

## Summary

Related meta issue: https://github.com/elastic/kibana/issues/131551

In this PR support is added for selecting/highlighting an area in one
scatterplot chart highlighting the corresponding area in the rest of the
charts.

- adds `user_selection` to spec
- adds `user_selection` condition in `encoding` `color` to ensure
highlighting in corresponding area of other charts
- outlier and user selection (by drag or single point click) are now
different colors


#### Classification results:

<img width="929" alt="image"
src="https://user-images.githubusercontent.com/6446462/200012485-18cee3f9-8180-4727-894c-6d8376678c03.png">

#### Classification job wizard:

<img width="928" alt="image"
src="https://user-images.githubusercontent.com/6446462/200013588-2445f475-274c-4548-bcc3-4af4d631a6c4.png">

#### Outlier Detections results:

<img width="969" alt="image"
src="https://user-images.githubusercontent.com/6446462/200036149-4491ea6e-77a5-4c1b-9ddc-386d8a492f69.png">

#### Outlier Detection job wizard:

<img width="938" alt="image"
src="https://user-images.githubusercontent.com/6446462/200036269-b3b8ce39-6956-4bf9-b7ff-f61eb2022ae8.png">

#### Regression results:

<img width="922" alt="image"
src="https://user-images.githubusercontent.com/6446462/200013085-248eff9a-09f5-42a3-ac11-c2bfc3c08cf4.png">

#### Regression job wizard:

<img width="944" alt="image"
src="https://user-images.githubusercontent.com/6446462/200013485-0c54808c-2cc2-44e6-bc6d-bb1c3825cbe8.png">

### Checklist

Delete any items that are not applicable to this PR.

- [ ] Any text added follows [EUI's writing
guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses
sentence case text and includes [i18n
support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [ ]
[Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html)
was added for features that require explanation or tutorials
- [x] [Unit or functional
tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html)
were updated or added to match the most common scenarios
- [ ] Any UI touched in this PR is usable by keyboard only (learn more
about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [ ] Any UI touched in this PR does not create any new axe failures
(run axe in browser:
[FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/),
[Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [ ] If a plugin configuration key changed, check if it needs to be
allowlisted in the cloud and added to the [docker
list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [ ] This renders correctly on smaller devices using a responsive
layout. (You can test this [in your
browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [x] This was checked for [cross-browser
compatibility](https://www.elastic.co/support/matrix#matrix_browsers)

Co-authored-by: kibanamachine <42973632+kibanamachine@users.noreply.github.com>

walterra committed:
[ML] Explain Log Rate Spikes: Apply random sampling to histogram aggregations. (#144627)

Adds random sampling as an option to `fetchFieldHistograms()`. Since this already supported normal sampling, this also adds a check if only one of the two options is used. Pending a future update we might remove support for normal sampling in a follow up. The Explain Log Rate Spikes analysis makes use of the new option and applies random sampling for the mini histogram data it fetches.

PhilippeOberti committed:
[TIP] Fixing row height on indicators data grid (#144290)

https://github.com/elastic/security-team/issues/5200

nchaulet committed:
[Fleet] Fix Fleet server hosts UI after design review (#144734)

maxcold committed:
[TIP] add field existence check to the painless script (#144344)

- fix fields in RawIndicatorFieldId (for investigate in timeline hook crash)
- fix logic to check if a field exist in the painless script
- add more e2e to test with missing fields and also missing mappings file

https://github.com/elastic/security-team/issues/5296

hop-dev committed:
[Fleet] Check cache for packageInfo before downloading archive (#144788)

seanstory committed:
Add features mapping for connectors (#144748)

## Summary

Part of https://github.com/elastic/enterprise-search-team/issues/3254

this augments the mappings for `.elastic-connectors` to include the
`features` field

#### Related PRs
- https://github.com/elastic/ent-search/pull/7027
- https://github.com/elastic/connectors-ruby/pull/408

Commit:
merge main

Commit:
add callout for no is one of template with preventing safe

Commit:
Merge branch '142435-add-is-one-of-operator-for-data-providers' of github.com:jamster10/kibana into 142435-add-is-one-of-operator-for-data-providers

jamster10 commented:
disregard review request. I merged main into my branch and pushed. not sure what happened there :/ Apologies for all the accidental pings working on removing requests soon!

jamster10 commented:
Closed in favour of https://github.com/elastic/kibana/pull/144988
